File: branches/0.9.6-glade/bcore/status.cpp
/*
 * btg Copyright (C) 2005 Michael Wojciechowski.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*
 * $Id$
 */

#include "status.h"
#include <bcore/t_string.h>
#include <bcore/hru.h>
#include <bcore/hrr.h>

#include <bcore/helpermacro.h>

namespace btg
{
   namespace core
   {

      using namespace std;

      Status::Status()
         : Serializable(),
           Printable(),
           contextID_(-1),
           filename_("undefined"),
           status_(ts_undefined),
           downloadTotal_(0),
           uploadTotal_(0),
           failedBytes_(0),
           downloadRate_(0),
           uploadRate_(0),
           done_(0),
           filesize_(0),
           leechers_(0),
           seeders_(0),
           time_left_d_(0),
           time_left_h_(0),
           time_left_m_(0),
           time_left_s_(0),
           trackerStatus_(),
	   activityCounter_(0)
      {}

      Status::Status(t_int const _contextID,
                     string const& _filename,
                     torrent_status const _status,
                     t_ulong const _dn_total,
                     t_ulong const _ul_total,
                     t_ulong const _failed_bytes,
                     t_ulong const _dn_rate,
                     t_ulong const _ul_rate,
                     t_int const _done,
                     t_ulong const _filesize,
                     t_int const _leechers,
                     t_int const _seeders,
                     t_int const _time_left_d,
                     t_ulong const _time_left_h,
                     t_int const _time_left_m,
                     t_int _time_left_s,
                     btg::core::trackerStatus _trackerStatus,
		     t_ulong const _activityCounter)
         : Serializable(),
           Printable(),
           contextID_(_contextID),
           filename_(_filename),
           status_(_status),
           downloadTotal_(_dn_total),
           uploadTotal_(_ul_total),
           failedBytes_(_failed_bytes),
           downloadRate_(_dn_rate),
           uploadRate_(_ul_rate),
           done_(_done),
           filesize_(_filesize),
           leechers_(_leechers),
           seeders_(_seeders),
           time_left_d_(_time_left_d),
           time_left_h_(_time_left_h),
           time_left_m_(_time_left_m),
           time_left_s_(_time_left_s),
           trackerStatus_(_trackerStatus),
	   activityCounter_(_activityCounter)
      {}

      Status::Status(Status const& _status)
         : Serializable(),
           Printable(),
           contextID_(_status.contextID_),
           filename_(_status.filename_),
           status_(_status.status_),
           downloadTotal_(_status.downloadTotal_),
           uploadTotal_(_status.uploadTotal_),
           failedBytes_(_status.failedBytes_),
           downloadRate_(_status.downloadRate_),
           uploadRate_(_status.uploadRate_),
           done_(_status.done_),
           filesize_(_status.filesize_),
           leechers_(_status.leechers_),
           seeders_(_status.seeders_),
           time_left_d_(_status.time_left_d_),
           time_left_h_(_status.time_left_h_),
           time_left_m_(_status.time_left_m_),
           time_left_s_(_status.time_left_s_),
           trackerStatus_(_status.trackerStatus_),
	   activityCounter_(_status.activityCounter_)
      {}

      void Status::set(t_int const _contextID,
                       string const& _filename,
                       torrent_status const _status,
                       t_ulong const _dn_total,
                       t_ulong const _ul_total,
                       t_ulong const _failed_bytes,
                       t_ulong const _dn_rate,
                       t_ulong const _ul_rate,
                       t_int const _done,
                       t_ulong const _filesize,
                       t_int const _leechers,
                       t_int const _seeders,
                       t_int const _time_left_d,
                       t_ulong const _time_left_h,
                       t_int const _time_left_m,
                       t_int _time_left_s,
                       btg::core::trackerStatus _trackerStatus,
		       t_ulong const _activityCounter)
      {
         contextID_         = _contextID;
         filename_          = _filename;
         status_            = _status;
         downloadTotal_     = _dn_total;
         uploadTotal_       = _ul_total;
         failedBytes_       = _failed_bytes;
         downloadRate_      = _dn_rate;
         uploadRate_        = _ul_rate;
         done_              = _done;
         filesize_          = _filesize;
         leechers_          = _leechers;
         seeders_           = _seeders;
         time_left_d_       = _time_left_d;
         time_left_h_       = _time_left_h;
         time_left_m_       = _time_left_m;
         time_left_s_       = _time_left_s;
         trackerStatus_     = _trackerStatus;
	 activityCounter_   = _activityCounter;
      }

      bool Status::serialize(btg::core::externalization::Externalization* _e) const
      {
         // int, context id;
         t_int temp_contextid  = this->contextID_;
         BTG_RCHECK( _e->intToBytes(&temp_contextid) );

         // string, filename;
         BTG_RCHECK( _e->stringToBytes(&this->filename_) );

         // int, status;
         t_int temp_status  = this->status_;
         BTG_RCHECK( _e->intToBytes(&temp_status) );

         //long, dn_total;
         BTG_RCHECK( _e->uLongToBytes(&this->downloadTotal_) );

         //long, ul_total;
         BTG_RCHECK( _e->uLongToBytes(&this->uploadTotal_) );

         //long, failed_bytes;
         BTG_RCHECK( _e->uLongToBytes(&this->failedBytes_) );

         //long, dn_rate;
         BTG_RCHECK( _e->uLongToBytes(&this->downloadRate_) );

         //long, ul_rate;
         BTG_RCHECK( _e->uLongToBytes(&this->uploadRate_) );

         //int, done;
         BTG_RCHECK( _e->intToBytes(&this->done_) );

         //long, filesize;
         BTG_RCHECK( _e->uLongToBytes(&this->filesize_) );

         //int, leeches;
         BTG_RCHECK( _e->intToBytes(&this->leechers_) );

         //int, seeders;
         BTG_RCHECK( _e->intToBytes(&this->seeders_) );

         //long, time left, days;
         BTG_RCHECK( _e->intToBytes(&this->time_left_d_) );

         //long, time left, hours;
         BTG_RCHECK( _e->uLongToBytes(&this->time_left_h_) );

         //int, time left, minutes;
         BTG_RCHECK( _e->intToBytes(&this->time_left_m_) );

         //int, time left, seconds;
         BTG_RCHECK( _e->intToBytes(&this->time_left_s_) );

         //tracker status;
         BTG_RCHECK( trackerStatus_.serialize(_e) );

	 // unsigned long, counter, minutes;
         BTG_RCHECK( _e->uLongToBytes(&this->activityCounter_) );

         return true;
      }

      bool Status::deserialize(btg::core::externalization::Externalization* _e)
      {
         // int, context id;
         t_int temp_contextid  = 0;
         BTG_RCHECK( _e->bytesToInt(&temp_contextid) );

         this->contextID_ = temp_contextid;

         // string, filename;
         BTG_RCHECK( _e->bytesToString(&this->filename_) );

         // int, status;
         t_int temp_status  = 0;
         BTG_RCHECK( _e->bytesToInt(&temp_status) );

         // Check that the input integer is a status.
         switch(temp_status)
            {
            case Status::ts_undefined:
            case Status::ts_queued:
            case Status::ts_checking:
            case Status::ts_connecting:
            case Status::ts_downloading:
            case Status::ts_seeding:
            case Status::ts_stopped:
            case Status::ts_finished:
               break;
            default:
               BTG_RCHECK(false);
            }

         this->status_ = static_cast<torrent_status>(temp_status);

         //long, dn_total;
         BTG_RCHECK( _e->bytesToULong(&this->downloadTotal_) );

         //long, ul_total;
         BTG_RCHECK( _e->bytesToULong(&this->uploadTotal_) );

         //long, failed_bytes;
         BTG_RCHECK( _e->bytesToULong(&this->failedBytes_) );

         //long, dn_rate;
         BTG_RCHECK( _e->bytesToULong(&this->downloadRate_) );

         //long, ul_rate;
         BTG_RCHECK( _e->bytesToULong(&this->uploadRate_) );

         //int, done;
         BTG_RCHECK( _e->bytesToInt(&this->done_) );

         //long, filesize;
         BTG_RCHECK( _e->bytesToULong(&this->filesize_) );

         //int, leeches;
         BTG_RCHECK( _e->bytesToInt(&this->leechers_) );

         //int, seeders;
         BTG_RCHECK( _e->bytesToInt(&this->seeders_) );

         //int, time left, days;
         BTG_RCHECK( _e->bytesToInt(&this->time_left_d_) );

         //long, time left, hours;
         BTG_RCHECK( _e->bytesToULong(&this->time_left_h_) );

         //int, time left, minutes;
         BTG_RCHECK( _e->bytesToInt(&this->time_left_m_) );

         //int, time left, seconds;
         BTG_RCHECK( _e->bytesToInt(&this->time_left_s_) );

         // Trackerstatus
         BTG_RCHECK( trackerStatus_.deserialize(_e));

	 // unsigned long, counter, minutes;
         BTG_RCHECK( _e->bytesToULong(&this->activityCounter_) );

         return true;
      }

      string Status::toString() const
      {
         string output("");

         output += "contex id: "    + convertToString<t_int>(contextID_) + ", ";
         output += "filename: '"    + filename_ + "', ";
         output += "status: "       + convertToString<t_int>(status_) + ", ";
         output += "trackerstatus: "+ trackerStatus_.toString() + ", ";
         output += "dn_total: "     + convertToString<t_ulong>(downloadTotal_) + ", ";
         output += "ul_total: "     + convertToString<t_ulong>(uploadTotal_) + ", ";
         output += "failed_bytes: " + convertToString<t_ulong>(failedBytes_) + ", ";
         output += "dn_rate: "      + convertToString<t_ulong>(downloadRate_) + ", ";
         output += "ul_rate: "      + convertToString<t_ulong>(uploadRate_) + ", ";
         output += "done: "         + convertToString<t_int>(done_) + ", ";
         output += "filesize: "     + convertToString<t_ulong>(filesize_) + ", ";
         output += "leechers: "      + convertToString<t_int>(leechers_) + ", ";
         output += "seeders: "      + convertToString<t_int>(seeders_) + ", ";

         output += "progress: "     + convertToString<t_int>(time_left_d_) + ":" + convertToString<t_ulong>(time_left_h_) + ":" +
            convertToString<t_int>(time_left_m_) + ":" +
            convertToString<t_int>(time_left_s_) + ", ";

	 output += "activity counter:" + convertToString<t_ulong>(activityCounter_);

         return output;
      }

      bool Status::operator== (Status const& _status) const
      {
         bool op_status = true;

         if ((this->contextID_ != _status.contextID_) ||
	     (this->filename_ != _status.filename_) || 
	     (this->status_ != _status.status_) || 
	     (this->downloadTotal_ != _status.downloadTotal_) || 
	     (this->uploadTotal_ != _status.uploadTotal_) || 
	     (this->failedBytes_ != _status.failedBytes_) || 
	     (this->downloadRate_ != _status.downloadRate_) || 
	     (this->uploadRate_ != _status.uploadRate_) ||
	     (this->done_ != _status.done_) ||
	     (this->filesize_ != _status.filesize_) ||
	     (this->leechers_ != _status.leechers_) ||
	     (this->seeders_ != _status.seeders_) ||
	     (this->time_left_d_ != _status.time_left_d_) ||
	     (this->time_left_h_ != _status.time_left_h_) ||
	     (this->time_left_m_ != _status.time_left_m_) ||
	     (this->time_left_s_ != _status.time_left_s_) ||
	     (this->trackerStatus_ != _status.trackerStatus_) ||
	     (activityCounter_ != _status.activityCounter_))
	   {
	     op_status = false;
	   }

         return op_status;
      }

      Status& Status::operator= (Status const& _status)
      {
         bool eq = (_status == *this);

         if (!eq)
            {
               contextID_      = _status.contextID_;
               filename_       = _status.filename_;
               status_         = _status.status_;
               downloadTotal_  = _status.downloadTotal_;
               uploadTotal_    = _status.uploadTotal_;
               failedBytes_    = _status.failedBytes_;
               downloadRate_   = _status.downloadRate_;
               uploadRate_     = _status.uploadRate_;
               done_           = _status.done_;
               filesize_       = _status.filesize_;
               leechers_       = _status.leechers_;
               seeders_        = _status.seeders_;
               time_left_d_    = _status.time_left_d_;
               time_left_h_    = _status.time_left_h_;
               time_left_m_    = _status.time_left_m_;
               time_left_s_    = _status.time_left_s_;
               trackerStatus_   = _status.trackerStatus_;
	       activityCounter_ = _status.activityCounter_;
            }

         return *this;
      }

     bool Status::validTimeLeft() const
     {
       bool status = false;

       if ((time_left_d_ != 0) || 
	   (time_left_h_ != 0) ||
	   (time_left_m_ != 0) || 
	   (time_left_s_ != 0))
	 {
	   status = true;
	 }

       return status;
     }

     void Status::timeLeftToString(std::string & _output, 
				   bool const _shortFormat) const
     {
       if (time_left_d_ > 0)
	 {
	   // Show only days.
	   _output += convertToString<t_ulong>(time_left_d_);

	   if (!_shortFormat)
	     {
	       if (time_left_d_ > 1)
		 {
		   _output += " days";
		 }
	       else
		 {
		   _output += " day";
		 }
	     }
	   else
	     {
	       _output += " d";
	     }
	 }
       else
	 {
	   // Show only h:m:s.

	   if (time_left_h_ < 10)
	     {
	       _output += "0";
	     }
	   _output += convertToString<t_ulong>(time_left_h_) + ":";

	   if (time_left_m_ < 10)
	     {
	       _output += "0";
	     }
	   _output += convertToString<t_int>(time_left_m_) + ":";
	   
	   if (time_left_s_ < 10)
	     {
	       _output += "0";
	     }
	   _output += convertToString<t_int>(time_left_s_);
	 }	  
     }

      Status::~Status()
      {}

   } // namespace core
} // namespace btg

File: tags/1.0/UI/www/php/htdocs/bcore/status.php
<?php
/*
 * btg Copyright (C) 2005 Michael Wojciechowski.
 * PHP client written by Johan Ström.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*
 * $Id: status.php,v 1.1.2.5 2007/02/09 07:33:30 jstrom Exp $
 */
if(!defined("BTG_BCORE_PATH"))
	define("BTG_BCORE_PATH", "bcore");

require_once(BTG_BCORE_PATH."/btgexception.php");
require_once(BTG_BCORE_PATH."/serializable.php");
require_once(BTG_BCORE_PATH."/trackerstatus.php");

class Status extends BTGSerializable
{
	const ts_undefined = 0;		// Not defined
	const ts_queued = 1;			// In queue for downloading.
	const ts_checking	= 2;		// The client is hash-checking the data.
	const ts_connecting = 3;	// The client is connecting.
	const ts_downloading = 4;	// The client is downloading.
	const ts_seeding = 5;		// The client is seeding. 
	const ts_stopped = 6;		// The client stopped downloading
	const ts_finished = 7;		// The client finished downloading and is not seeding.

	private $statusNames = array(
		"Undefined", 
		"Queued",
		"Checking data",
		"Connecting",
		"Downloading",
		"Seeding",
		"Stopped",
		"Finished",
		);
	
	/// The context ID
	private $contextID;
	/// The filename
	private $filename;
	/// The status, should be one of the above ts_* consts.
	private $status;
	/// Total downloaded bytes
	private $downloadTotal;
	/// Total uploaded bytes
	private $uploadTotal;
	/// Failed bytes
	private $failedBytes;
	/// Download rate
	private $downloadRate;
	/// Upload rate
	private $uploadRate;
	/// Done in percent
	private $done;
	/// Total filesize in bytes
	private $filesize;
	/// Number of leechers.
	private $leechers;
	/// Number of seeders
	private $seeders;

	// Time left, in days hours minutes and seconds
	private $time_left_d;
	private $time_left_h;
	private $time_left_m;
	private $time_left_s;

	// Trackerstatus
	private $trackerStatus;

	// Activity counter
	private $activityCounter;

	private $hash;

        private $url;

	public function Status($contextID = -1, $filename="undefined", $status=0, $dn_total=0, $ul_total=0, $failed_bytes=0, $dn_rate=0, $ul_rate=0, $done=0, $filesize=0, $leechers=0, $seeders=0, $time_left_d=0, $time_left_h=0, $time_left_m=0, $time_left_s=0, $trackerStatus = null, $activityCounter = 0, $hash = "", $url = "")
	{
		if($trackerStatus == null)
				$trackerStatus = new trackerStatus();
		$this->set($contextID, $filename, $status, $dn_total, $ul_total, $failed_bytes, $dn_rate, $ul_rate, $done, $filesize, $leechers, $seeders, $time_left_d, $time_left_h, $time_left_m, $time_left_s, $trackerStatus, $activityCounter, $hash, $url);
	}

	public function set($contextID, $filename, $status, $dn_total, $ul_total, $failed_bytes, $dn_rate, $ul_rate, $done, $filesize, $leechers, $seeders, $time_left_d, $time_left_h, $time_left_m, $time_left_s, $trackerStatus, $activityCounter, $hash, $url)
	{
		$this->contextID = $contextID;
		$this->filename = $filename;
		$this->status = $status;
		$this->downloadTotal = $dn_total;
		$this->uploadTotal = $ul_total;
		$this->failedBytes = $failed_bytes;
		$this->downloadRate = $dn_rate;
		$this->uploadRate = $ul_rate;
		$this->done = $done;
		$this->filesize = $filesize;
		$this->leechers = $leechers;
		$this->seeders = $seeders;
		$this->time_left_d = $time_left_d;
		$this->time_left_h = $time_left_h;
		$this->time_left_m = $time_left_m;
		$this->time_left_s = $time_left_s;
		$this->trackerStatus = $trackerStatus;
		$this->activityCounter = $activityCounter;
		$this->hash            = $hash;
		$this->url             = $url;
	}

	function getContextID() { return $this->contextID; }
	function getFilename()	{ return $this->filename; }
	function getStatus()	{ return $this->status; }
	function getStatusName()	{ return $this->statusNames[$this->status]; }
	function getDownloadTotal()	{ return $this->downloadTotal; }
	function getUploadTotal()	{ return $this->uploadTotal; }
	function getfailedBytes()	{ return $this->failedBytes; }
	function getDownloadRate()	{ return $this->downloadRate; }
	function getUploadRate()	{ return $this->uploadRate; }
	function getDone()	{ return $this->done; }
	function getFilesize()	{ return $this->filesize; }
	function getLeechers()	{ return $this->leechers; }
	function getSeeders()	{ return $this->seeders; }
	function getTimeLeftD()	{ return $this->time_left_d; }
	function getTimeLeftH()	{ return $this->time_left_h; }
	function getTimeLeftM()	{ return $this->time_left_m; }
	function getTimeLeftS()	{ return $this->time_left_s; }
	function getTrackerStatus()	{ return $this->trackerStatus; }
	function getActivityCounter()	{ return $this->activityCounter; }

	function getHash() { return $this->hash; }
   function getAnnounceURL() { return $this->url; }

	public function serialize(&$a = array())
	{
		$this->intToBytes($a, $this->contextID);
		$this->stringToBytes($a, $this->filename);
		$this->intToBytes($a, $this->status);
		$this->uLongToBytes($a, $this->downloadTotal);
		$this->uLongToBytes($a, $this->uploadTotal);
		$this->uLongToBytes($a, $this->failedBytes);
		$this->uLongToBytes($a, $this->downloadRate);
		$this->uLongToBytes($a, $this->uploadRate);
		$this->intToBytes($a, $this->done);
		$this->uLongToBytes($a, $this->filesize);
		$this->intToBytes($a, $this->leechers);
		$this->intToBytes($a, $this->seeders);
		$this->intToBytes($a, $this->time_left_d);
		$this->uLongToBytes($a, $this->time_left_h);
		$this->intToBytes($a, $this->time_left_m);
		$this->intToBytes($a, $this->time_left_s);
		$this->trackerStatus->serialize($a);
		$this->uLongToBytes($a, $this->activityCounter);
		$this->stringToBytes($a, $this->hash);
		$this->stringToBytes($a, $this->url);
		return $a;
	}

	public function deserialize(&$data)
	{
		$this->bytesToInt($this->contextID, $data);
		$this->bytesToString($this->filename, $data);

		$temp_status = 0;
		$this->bytesToInt($temp_status, $data);

		switch($temp_status)
		{
			case Status::ts_undefined:
			case Status::ts_queued:
			case Status::ts_checking:
			case Status::ts_connecting:
			case Status::ts_downloading:
			case Status::ts_seeding:
			case Status::ts_stopped:
			case Status::ts_finished:
				break;
			default:
				throw new BTGException("Status::deserialize failed to deserialize status. Invalid status '$temp_status'.");
		}
		$this->status = $temp_status;

		$this->bytesToULong($this->downloadTotal, $data);
		$this->bytesToULong($this->uploadTotal, $data);
		$this->bytesToULong($this->failedBytes, $data);
		$this->bytesToULong($this->downloadRate, $data);
		$this->bytesToULong($this->uploadRate, $data);
		$this->bytesToInt($this->done, $data);
		$this->bytesToULong($this->filesize, $data);
		$this->bytesToInt($this->leechers, $data);
		$this->bytesToInt($this->seeders, $data);
		$this->bytesToInt($this->time_left_d, $data);
		$this->bytesToULong($this->time_left_h, $data);
		$this->bytesToInt($this->time_left_m, $data);
		$this->bytesToInt($this->time_left_s, $data);
		$this->trackerStatus->deserialize($data);
		$this->bytesToULong($this->activityCounter, $data);
		$this->bytesToString($this->hash, $data);
		$this->bytesToString($this->url, $data);
	}

	public function toString($lineWidth = 80)
	{
		$ret = "context id: ". $this->contextID . "\n";
		$ret.= "filename: ". $this->filename . "\n";
		$ret.= "status: ". $this->status . "\n";
		$ret.= "dn_total: ". $this->downloadTotal . "\n";
		$ret.= "ul_total: ". $this->uploadTotal . "\n";
		$ret.= "failed_bytes: ". $this->failedBytes . "\n";
		$ret.= "dn_rate: ". $this->downloadRate . "\n";
		$ret.= "ul_rate: ". $this->uploadRate . "\n";
		$ret.= "done: ". $this->done . "\n";
		$ret.= "filesize: ". $this->filesize . "\n";
		$ret.= "leechers: ". $this->leechers. "\n";
		$ret.= "seeders: ". $this->seeders . "\n";
		$ret.= "time left: ". $this->time_left_d. "d, ". $this->time_left_h . "h, " . $this->time_left_m . "m, " . $this->time_left_s . "s\n";
		$ret.= "trackerstatus: ".$this->trackerStatus->toString()."\n";
		$ret.= "activity counter: ".$this->activityCounter."\n";

		return $ret;
	}
}
?>
